### PR TITLE
Added new inputlocation for downloading jars from maven

### PR DESCRIPTION
**Pull request by palaniappan1**

### CODE REVIEW

File: sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        int bytesRead;
+        while ((bytesRead = inputStream.read(buffer, 0, BUFFER_SIZE)) != -1) {
+          outputStream.write(buffer, 0, bytesRead);
+        }
+      }
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    return file.exists() ? file.toPath() : null;
```
Comment (swissiety) [inline sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java]: .

**Review by swissiety — APPROVED** [review]

File: sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      String downloadURL, List<BodyInterceptor> bodyInterceptors, Collection<Path> ignoredPaths) {
+    super(
+        downloadAndConstructPath(downloadURL), SourceType.Library, bodyInterceptors, ignoredPaths);
+  }
+
+  private static Path downloadAndConstructPath(String downloadURL) {
+    HttpURLConnection connection;
+    String tempDirPath = System.getProperty("java.io.tmpdir");
+    String filename = downloadURL.substring(downloadURL.lastIndexOf("/") + 1);
+    File file = new File(tempDirPath, filename);
```
Comment (marcus-h) [inline sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java]: Hmm shouldn't we use a real tmpfile? Writing to arbitrary files in the tmpdir
is not so nice:)

File: sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      try (InputStream inputStream = new BufferedInputStream(connection.getInputStream());
+          OutputStream outputStream = Files.newOutputStream(Paths.get(file.getAbsolutePath()))) {
+        byte[] buffer = new byte[BUFFER_SIZE];
+        int bytesRead;
+        while ((bytesRead = inputStream.read(buffer, 0, BUFFER_SIZE)) != -1) {
+          outputStream.write(buffer, 0, bytesRead);
+        }
+      }
+    } catch (IOException e) {
+      throw new RuntimeException(e);
```
Comment (marcus-h) [inline sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java]: Hmm maybe an UncheckedIOException?

File: sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  private static Path downloadAndConstructPath(String downloadURL) {
+    HttpURLConnection connection;
+    String tempDirPath = System.getProperty("java.io.tmpdir");
+    String filename = downloadURL.substring(downloadURL.lastIndexOf("/") + 1);
+    File file = new File(tempDirPath, filename);
+    try {
+      URL url = new URL(downloadURL);
+      connection = (HttpURLConnection) url.openConnection();
```
Comment (marcus-h) [inline sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java:53]: Is this a potential resource leak? Do we have to manually close the connection? (I haven't checked the API...)

File: sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java (hunk 2 above)
Comment (palaniappan1) [reply sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java]: This is the real temp file, tmp directory by JVM.

File: sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java (hunk 3 above)
Comment (palaniappan1) [reply sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java]: I don't think Unchecked is really needed here. Simple file operations, so IOException itself is enough i think.

File: sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java (hunk 4 above)
Comment (palaniappan1) [reply sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java:53]: Good question. Fixed it!

File: sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java (hunk 2 above)
Comment (marcus-h) [reply sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java]: A user could have intentionally created a file called /tmp/foo-1.2.3.jar,
which is then overwritten by the downloader. IMHO, we shouldn't do that.
Let's use a "real" tmpfile, which is created, for instance, via
Files.createTempFile(...).

File: sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java (hunk 3 above)
Comment (marcus-h) [reply sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/DownloadJarAnalysisInputLocation.java]: I was talking about "new RuntimeException(e)" vs. "new UncheckedIOException(e)".

**Pull request merged**